pygmt/src/plot.py

```diff
@@ -218,6 +218,14 @@ def plot(self, x=None, y=None, data=None, sizes=None, direction=None, **kwargs):
             )
         extra_arrays.append(sizes)
 
+    if "I" in kwargs and is_nonstr_iter(kwargs["I"]):
+        if kind != "vectors":
+            raise GMTInvalidInput(
+                "Can't use arrays for intensity if data is matrix or file."
+            )
+        extra_arrays.append(kwargs["I"])
+        kwargs["I"] = ""
+
     if "t" in kwargs and is_nonstr_iter(kwargs["t"]):
         extra_arrays.append(kwargs["t"])
         kwargs["t"] = ""
```

Comment:
Two comments here:

- Many parameters (e.g., sizes, colors, transparency) don't work if data is matrix or file. I feel we should have a more general way to do the check, rather than raising the GMTInvalidInput exception for each parameter.

- The two options **-I** and **-t** share very similar codes, and we should be able to simplify it like:
```
for flag in ["I", "t"]:
    if flag in kwargs and is_nonstr_iter(kwargs[flag]):
        extra_arrays.append(kwargs[flag])
        kwargs[flag] = ""
```

Reply:
> * Many parameters (e.g., sizes, colors, transparency) don't work if data is matrix or file. I feel we should have a more general way to do the check, rather than raising the GMTInvalidInput exception for each parameter.

We can use the `columns` parameter ([-i](https://docs.generic-mapping-tools.org/latest/gmt.html#icols-full) in GMT) to select which columns in `data` to use, right? Is the reason we raise this error that we don't want to mix using array-like `x`, `y`, `sizes` etc with the `data` parameter because this will make PyGMT maintenance more difficult? We have to parse both the `data` parameter and those array-like parameters. Is my understanding right?

Reply:
> * The two options **-I** and **-t** share very similar codes, and we should be able to simplify it like:
> 
> ```
> for flag in ["I", "t"]:
>     if flag in kwargs and is_nonstr_iter(kwargs[flag]):
>         extra_arrays.append(kwargs[flag])
>         kwargs[flag] = ""
> ```

We don't need to raise an `GMTInvalidInput` for **t**.

Reply:
We need. The codes don't check it.

Reply:
An array-like **t** works for both `data` and `x`/`y`, right? Now we only need to raise `GMTInvalideInput` for `color`, `sizes` and `intensity`. See https://github.com/GenericMappingTools/pygmt/pull/1065#discussion_r596161372

Reply:
array-like t doesn't work for data.

Reply:
You are right. I use the example in [Points with varying transparency](https://www.pygmt.org/latest/gallery/symbols/points_transparency.html#sphx-glr-gallery-symbols-points-transparency-py) to test it.
```Python
import numpy as np
import pygmt

x = np.arange(0, 105, 5)
y = np.ones(x.size)
data = np.column_stack([x, y])
transparency = x

fig = pygmt.Figure()
fig.basemap(region=[-5, 105, 0, 2], frame=['xaf+l"Transparency"+u%', "WSrt"],projection="X15c/6c")
fig.plot(data=data, style="c0.6c", color="blue", pen="1p,red", transparency=transparency)
fig.show()
```
![trans test](https://user-images.githubusercontent.com/50591376/111868117-11758d80-89b3-11eb-8ed8-349f09825b29.png)